Use .ContinueWith instead of finally for removal.

tillig opened the pull request:
This allows non-async/await and Task-based handling to complete and still remove the scope without the try/finally removing it before the task is finished.

Based on comments for #17.

tillig commented:
@russellfoster @srogovtsev Give this a look, I think I got it. If you don't have time, no biggie, but thought you'd be interested. Plus, I have to admit my `Task` semantics are a little rusty since `async`/`await` came along. I had to come up with some pretty contrived examples to run tests on it.

alistairjevans commented on src/Autofac.Integration.WebApi.Owin/DependencyScopeHandler.cs:
```diff
@@ -42,13 +42,15 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
+            {
+                request.Properties.Remove(HttpPropertyKeys.DependencyScope);
+                return task.Result;
+            });
```
There's an analyser that indicates `ContinueWith` should always be used with an explicit scheduler? https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2008

tillig replied:
Weird, the analyzer didn't fire off and doesn't show in the build or VS. Let me add some of the stuff from the blog article they link in the doc there.

alistairjevans commented on test/Autofac.Integration.WebApi.Owin.Test/DependencyScopeHandlerFixture.cs:
```diff
@@ -64,14 +64,76 @@ public async void RemoveAutofacDependencyScopeFromHttpRequestMessageAfterLeaving
+    }
+
+    [Fact]
+    public void RemoveAutofacDependencyScopeEvenIfTaskThrows()
```
Is it worth having one of these tests as async (I think there's an `Assert.ThrowsAsync()`) to check the `ContinueWith` schedules correctly?

tillig replied:
Sure, I can add that.

alistairjevans reviewed (approved):
Approved with two questions. 👍

srogovtsev commented:
You'd also need a test for cancelled tasks, something like:

```cs
var cancellationTokenSource = new CancellationTokenSource();
cancellationTokenSource.Cancel();

var fakeHandler = new FakeInnerHandler(_ =>
        {
            return Task.Factory.StartNew<HttpResponseMessage>(async () =>
            {
                //this should cancel immediately
                await Task.Delay(-1, cancellationTokenSource.Token);
                //we never get here, but the compiler doesn't know it
               return new HttpResponseMessage(HttpStatusCode.OK);
            });
        });
```

srogovtsev commented on src/Autofac.Integration.WebApi.Owin/DependencyScopeHandler.cs:
```diff
@@ -42,13 +42,15 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
+        // marked code.
+        return base
+            .SendAsync(request, cancellationToken)
+            .ContinueWith(task =>
```
I'd probably use `.ContinueWith(..., TaskContinuationOptions.ExecuteSynchronously)`, as our continuation is extremely short, and we'll avoid any additional context switch.

tillig commented:
- Added continuation options to specify the scheduler and that the continuation should run synchronously.
- Added `async`/`await` test equivalents of the `Task` tests. This made the original test for scope removal redundant because the same scenario is being tested with the new tests.

srogovtsev commented on src/Autofac.Integration.WebApi.Owin/DependencyScopeHandler.cs:
```diff
@@ -42,13 +42,19 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
+                },
+                CancellationToken.None,
+                TaskContinuationOptions.ExecuteSynchronously,
+                TaskScheduler.FromCurrentSynchronizationContext());
```
Does `TaskScheduler.FromCurrentSynchronizationContext()` make sense when we use `TaskContinuationOptions.ExecuteSynchronously`?

(brief peek into `Task.ContinueWith` filled me with eldritch horrors)

tillig replied:
That's a good question. Reading the docs on the [analyzer](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2008) I got into [this blog article](https://devblogs.microsoft.com/pfxteam/new-taskcreationoptions-and-taskcontinuationoptions-in-net-4-5/) that seemed to recommend this as the way, but I suppose we could use `TaskScheduler.Current` to continue on the same thread, yeah? Again, my `Task` is a little rusty. ;)

srogovtsev replied:
I would probably just use the overload that needs only `TaskContinuationOptions` (it will still pass `TaskScheduler.Current`, but that will be out of our hands). I cannot check whether this will still trigger the analyzer, but if it will, I think we can safely suppress the warning saying "we don't care, really don't care where it runs".

PS I feel your pain, I haven't been manually constructing continuations since I don't know when.

tillig replied:
Ended up having to _add updated analyzers_ to see the analyzer fire... then found that it does get caught, so suppressed it. Which seems sort of weird but I'd rather know we thought about it and explicitly chose not to do anything rather than it seeming like we didn't think about it at all.

srogovtsev replied:
>  I'd rather know we thought about it and explicitly chose not to do anything rather than it seeming like we didn't think about it at all.

"Explicit is better than implicit".

alistairjevans replied:
Feel like a terrible person saying it, but isn't there an overload of `ContinueWith` that takes a CancellationToken? I think you need to pass the existing `cancellationToken` into the continuation to ensure the returned Task can get IsCanceled set correctly if the token gets cancelled.

Arguably it doesn't matter if the continuing action is short and synchronous, but I can't remember whether the original task's cancellation state is passed through to the returned Task.

Manual ContinueWith sucks.

tillig replied:
It seems like, from a `ContinueWith` overload perspective, you can either specify _almost nothing_ or _almost everything_ but not really both. What that means is:

- I added the original `cancellationToken` passing through, _but_
- I also now have to specify `TaskScheduler.Current` and can't leave it to defaulted, which means
- I don't have to suppress the analyzer anymore

If y'all could just move off old Web API and OWIN this wouldn't be an issue. 😆

srogovtsev replied:
> If y'all could just move off old Web API and OWIN this wouldn't be an issue.

If we could just move off _at least_ Web API, so much wouldn't be an issue...

But we, for one, do not require `[Security*]` markup, so we it for us, you'd be able to just do `await`.

tillig commented:
Pushing this through. I think I got everything and it should at least unblock folks who were affected by the issue. We can tweak which scheduler to use later if we need to.
